### PR TITLE
fix route config of context bug

### PR DESCRIPTION
**Pull request by runzx**
#### bug:
```js
const opt = {
  config: {
    auth: {
      scope: ['admin', 'super'],
      strategy: 'worker'
    }
  }
}
fastify.get('/test', opt, async (req, reply) => {}
opt.config.auth.strategy = 'simple'
fastify.get('/home', opt, async (req, reply) => {}
// these reply.context.config is same:
{
  ...
    strategy:"simple",
  url:"/home"
}
```

### CODE REVIEW

**Comment by mcollina** [comment]
Can you please add a unit test?

File: fastify.js [hunk 1]
```diff
@@ -548,13 +548,13 @@ function build (options) {
...
 
-    options = Object.assign({}, options, {
+    const opt = Object.assign({}, options, {
       method,
       url,
       handler: handler || (options && options.handler)
     })
-
-    return _fastify.route(options)
+    options.config && (opt.config = JSON.parse(JSON.stringify(options.config)))
```
Comment (mcollina) [inline fastify.js]: Can you please use https://github.com/davidmarkclements/rfdc instead? Functions are not preserved in this way, and a function _might_ be part of the config.

**Comment by runzx** [comment]
no unit test . 
I test in my project.

**Comment by mcollina** [comment]
It would be great to add a unit test in Fastify as well so that we do not accidentally regress on this.

**Comment by mcollina** [comment]
You'd need to add another test in https://github.com/fastify/fastify/blob/master/test/route.test.js.

**Comment by allevo** [comment]
I'm not so convinced that this is a bug.
fastify allow you to add the configuration to an handler. So the developer know that *that* object is shared.

Anyway this can increase the memory greatly. Imho if the developer want not to share the same reference,  can use the deepcopy in the project code.

NB: this can be consider a breaking change too.

**Comment by mcollina** [comment]
@allevo I'm also convinced of your point: not copying it enables more powerful behavior, such as shared caches. However if we decide not to change this, we surely need to document it.

**Comment by jsumners** [comment]
I don't even see why this needs to be documented, really. The original description clearly shows that the author is mutating the config object after supplying it to one route before supplying it to another route. What does one expect when they mutate the object?

**Comment by runzx** [comment]
> It would be great to add a unit test in Fastify as well so that we do not accidentally regress on this.

**Comment by runzx** [comment]
Sorry, I didn't explain clearly in the beginning. 

I think that the two router config should get the different return value. 
1. the first "strategy" should be "worker" and the second one is "simple"
2. the first "url" should return "/test" and the second one returns "/home"
But during my test in the project, both router config reply.context.config return same value: 
stragety:"simple"
url:"/home"

I suppose that the method "fastify.get" uses object "opt" without deep copy. It causes the wrong context for "reply.context.config".

**Comment by runzx** [comment]
I don't know how to use 'tap' unit test yet. 
I'm investigating it now and will submit my unit test later.

**Comment by mcollina** [comment]
Forcing a deep copy on the config object prevents developers to specific complex objects and data structures that might not be cloneable. If you want to reuse the same option object, you should be doing the copying yourself.

@jsumners I think we should just add some docs in 
https://www.fastify.io/docs/latest/Routes stating that the option object is not deep cloned.

**Comment by jsumners** [comment]
It can’t hurt to add the note to the docs. But I don’t know why one would expect otherwise. Objects are passed by reference in JS. This is not a failing of Fastify but a failure to think things through.

**Comment by runzx** [comment]
If we don't do any changes, when I need to deal the authorization in the lifecycle of "onRequest", how to pass the return value to the router handler method?

**Comment by mcollina** [comment]
How does that relate to this?

**Comment by mcollina** [comment]
Is somebody in @fastify/fastify that thinks this is worth doing? Otherwise I think we should close this.

**Comment by cemremengu** [comment]
Looks like the majority is 👎 so I think we should skip this one

**Pull request closed without merging**